Add multiple commands to also support LFS image creation

HHHartmann opened the pull request:
This PR adds a ```build``` and an ```lfs-image``` command to the docker image.
The default script merely prints a short help screen.

HHHartmann committed:
also Support LFS images on image

Have multiple scripts to support building and generating LFS images.

HHHartmann committed:
should be Lua files not lfs files

marcelstoer commented:
Woah Gregor, good stuff, thanks!

marcelstoer commented on .gitattributes:
```diff
@@ -0,0 +1,7 @@
+lfs-image text eol=lf
+
+# Declare win dump file that will always have CRLF line endings on checkout.
+read.me text eol=crlf
```
Hmm, why is that so i.e. why is that one different?

HHHartmann replied:
I thought it would be printed on WinX mainly which likes crlf better than linux style lf. 
But wasn't  to sure about that.

marcelstoer replied:
Good thinking, I never thought about this as I'm not on Windows. You obviously are?

> I thought it would be printed on WinX mainly

No idea if this assumption is correct or not but I understand your reasoning. I'm a proponent of "form follows function". So, the readme just has to be legible enough on all platforms so that users can easily extract the relevant information. This seems fulfilled here.

marcelstoer commented on cmd.sh:
```diff
@@ -2,91 +2,4 @@
-if [ "$CAN_MODIFY_VERSION" = true ]; then
-  git checkout app/include/user_version.h
-fi
+cat ../read.me
```
Great you introduced the readme! However, I'm reluctant to change the command and thus give up backwards compatibility. Every script that's calling the image will break.

It may very well be a matter of personal taste and style but I'm currently leaning towards passing a param for the LFS stuff. I do acknowledge though that your proposal has its charm as it's much more  Dockerish to treat the two operations as explicit command.

HHHartmann replied:
hmm, also could leave the original functionality in place and only have an extra command to create an LFS image.
Not sure whether it makes sense to print the read.me at the end of the build then.

BTW: the build script has a weakness in building Float only builds. When int builds are configured in the config file it still builds an integer and names it float.

marcelstoer replied:
I'll sleep over that design decision.

> When int builds are configured in the config file it still builds an integer and names it float.

That recently surfaced somewhere in a comment. We should get rid of those INTEGER_ONLY/FLOAT_ONLY params and only honour the configuration in the `.h` file.

HHHartmann replied:
Sounds good :-)

marcelstoer replied:
If we're sacrificing backwards compatibility for this then we might as well break the INTEGER_ONLY/FLOAT_ONLY thing with it. Are you up for this?

So, instead of evaluating those params in `build` and `lfs-image` you would evaluate `#define LUA_NUMBER_INTEGRAL` in `user_config.h`. The documentation in the README could borrow the section from https://nodemcu.readthedocs.io/en/latest/en/build/#integer-build to explain.

HHHartmann replied:
Will add the new functionality in the next days. The developer will want to tweak the list of modules anyway, so why not adapt integer and LFS settings as well.

marcelstoer replied:
Just coming back to this and wondering if you'll have anything ready soon? Would be a pity if we both spent time for the same stuff. Want to get this done and over with asap so we can proceed with LFS merging in NodeMCU.

HHHartmann replied:
It is ready except for documentation and some testing.
Will be ready in about 2 Days.
Hope that is early enough so you can wait.

marcelstoer replied:
Great, looking forward, thanks.

HHHartmann replied:
@marcelstoer please have another look.

marcelstoer commented on lfs-image:
```diff
@@ -0,0 +1,36 @@
+cd /opt/nodemcu-firmware
+
+if [ ! -x luac.cross -a !  -x luac.cross.int ]; then
+  echo Error: No cross compiler found. You need to build the firmeware first.
```
I've been away from the LFS/luac.cross developments for a while as you certainly noticed. Did you recently change the firmware build so that `luac.cross` gets built with it automatically? When I last checked you needed to `make` it separately.

HHHartmann replied:
You can make it separately if you don't have/want the xtensa toolchain, but it also gets built in a firmware build.

marcelstoer commented on lfs-image:
```diff
@@ -0,0 +1,36 @@
+# make a float LFS image if available
+if [ -x luac.cross ]; then
+  echo creating "${IMAGE_NAME}".bin
+  find ../lua -iname "*.lua" -exec ./luac.cross -f -o ../lua/"${IMAGE_NAME}".bin {} +
```
Did you (i.e. Terry and you) recently change the file ending convention from `.img` to `.bin`?

HHHartmann replied:
Good catch. This should be img.

HHHartmann committed:
generate *.img files instead of *.bin

HHHartmann committed:
Merge branch 'master' into multi-command

HHHartmann committed:
remove INTEGER_ONLY and FLOAT_ONLY options

remove the options since they did not reliably work if LUA_NUMBER_INTEGRAL was defined in source.
Also add LFS state and Int/Float build type to Version info.

HHHartmann committed:
Merge branch 'multi-command' of https://github.com/HHHartmann/docker-nodemcu-build into multi-command

marcelstoer commented on build:
```diff
@@ -57,34 +72,22 @@ cd /opt/nodemcu-firmware
-else
-  true
-fi
+echo $BUILD_TYPE $CAN_MODIFY_VERSION
```
Left-over from your testing?

HHHartmann replied:
Yes, I saw it also.

marcelstoer commented on build:
```diff
@@ -57,34 +72,22 @@ cd /opt/nodemcu-firmware
 if [ "$CAN_MODIFY_VERSION" = true ]; then
-  MODULES=$(awk '/^[ \t]*#define LUA_USE_MODULES/{modules=modules sep tolower(substr($2,17));sep=","}END{if(length(modules)==0)modules="-";print modules}' app/include/user_modules.h)
-  sed -i '/#define NODE_VERSION[[:space:]]/ s/$/ " built with Docker provided by frightanic.com\\n\\tbranch: '"$BRANCH"'\\n\\tcommit: '"$COMMIT_ID"'\\n\\tSSL: '"$SSL"'\\n\\tmodules: '"$MODULES"'\\n"/g' app/include/user_version.h
+  sed -i 's;\r;;g' app/include/user_version.h
```
I think this fiddling with `\r` warrants a comment, no?

HHHartmann replied:
Just thought about macs, They have again different lineneds. So this might not work there and break the build.

marcelstoer replied:
I tested your previous version on my Mac and it was just fine IIRC.

marcelstoer commented on README.md:
```diff
@@ -29,26 +31,21 @@ Docker runs on a VirtualBox VM which by default only shares the user directory f
+Start Docker and change to the NodeMCU firmware directory (in the Docker console). To build the firmware run:
 
-`docker run --rm -it -v //c/Users/<user>/<nodemcu-firmware>:/opt/nodemcu-firmware marcelstoer/nodemcu-build`
+``docker run --rm -ti -v `pwd`:/opt/nodemcu-firmware marcelstoer/nodemcu-build build``
```
I guess you can confirm that Docker for Windows now properly supports `pwd`?

marcelstoer commented:
Great, Gregor! Almost ready to merge. There a one or two things in the README that I like to describe differently. I'll do that after the merge.

HHHartmann committed:
Add comment and remove testing code

marcelstoer commented:
Good job Gregor, thanks again. I now gave you proper credit in the README.

HHHartmann commented:
Marcel you are welcome and thanks for the credits!
